### DISCUSSION TITLE
Is there the way to convert remark nodes to amp components?

### DISCUSSION THREAD

**Discussion by illumination-k** [mdx-js/mdx#1898] [Q&A] [answered]
I try to make the SSG site with AMP and Next.js.

I want to convert remark node (e.g., image, math) to amp components (e.g., amp-img, amp-mathml) with custom remark extensions.

I tried to convert the remark node to `mdxFlowElement` with attributes like the following when compiling MDX, but I got the error `SyntaxError: Unexpected token '-'`.

```js
const newNode = {
  type: "mdxJsxFlowElement",
  name: "amp-mathml",
  attributes: [
    {type: "mdxJsxAttribute", name: "data-formula", value: `[${node.value}]`}
  ]
}
```

But compile successed with following MDX 
```
##h2

<amp-img
  src="https://amp.dev/static/inline-examples/images/mountains.webp?width=1750&hash=3ba88bf02bbdf1063bfdefa84c1955060b3a5fa6" 
  width="550" height="368"
/>
```

So is there a way to convert remark nodes to amp components? and how?

**Comment by wooorm** [comment, marked as answer]
Yes, you can turn mdast into several things, which includes finally getting AMP out.

Your question does not have enough information. Follow [Support](https://mdxjs.com/community/support/)

**Reply by illumination-k** [reply]
Thanks for your fast reply and sorry for the lack of information.

I created codesandbox,

https://codesandbox.io/s/exciting-butterfly-hk3r7?file=/pages/index.js

I would like to convert

```
## TL;DR
$$\sum$$
```

to

```
<h2>TL;DR</h2>
<amp-mathml data-formula=[\sum] />
```

in serverside and run it in client side.

**Reply by wooorm** [reply]
There is a `math` and an `inlineMath` node. There is an `mdxJsxFlowElement` and an `mdxJsxTextElement`. You’re only handling two.

**Reply by illumination-k** [reply]
Sorry, maybe I can't fully understand your message. 

I put the script to check the compile results in [gist](https://gist.github.com/illumination-k/6dc43d1b3391c429efb0b80ef919f752).

If I use `math` or `inlineMath` and `mdxJsxFlowElement` or `mdxJsxTextElement`, the compile result from `$$\sum$$` is

```
_jsx(_components.p, {
  children: _jsx(_components["amp-mathml"], {
    "data-formula": "[\sum]"
})
``` 

and error occured in `runSync(code, runtime)` with `SyntaxError: Unexpected token '-'`.

On the other hand, the compile result from
```mdx
<amp-img
  src="https://amp.dev/static/inline-examples/images/mountains.webp?width=1750&hash=3ba88bf02bbdf1063bfdefa84c1955060b3a5fa6"
  width="550"
  height="368"
/>
```

is 

```
_jsx("amp-img", {
  src: "https://amp.dev/static/inline-examples/images/mountains.webp?width=1750&hash=3ba88bf02bbdf1063bfdefa84c1955060b3a5fa6",
  width: "550",
  height: "368"
})
```

and `runSync(code, runtime)` works well.

`math` and `img` are just the examples. I would like to know how to convert remark node to `_jsx("amp-xxx, {...})` or how do I get `_jsx(_components["amp-xxx"], {...})` to work well in `runSync`.

**Reply by wooorm** [reply]
The syntax error you mentions is already being discussed/worked on/landing in https://github.com/remarkjs/remark-math/issues/72, https://github.com/wooorm/xdm/commit/ddcebdbcb577015a8930b310ee7a3ba045f1c24a, and https://github.com/wooorm/xdm/pull/106

**Reply by illumination-k** [reply]
Awesome! I'm really looking forward to the release. Thanks for your help.
